docs/git/cheat-sheet.md

```diff
@@ -0,0 +1,67 @@
+# Git Cheat Sheet
+
+The following is a cheat sheet for using [git] while developing Cardano SL
+which should contain all the git operations that someone working on Cardano
+will need at some point.
+
+### Cloning the Cardano SL repository
+```
+git clone https://github.com/input-output-hk/cardano-sl
+```
+After you change directory into that cloned repository, you may want to tell
+git to use the SSH protocol for the "git push" command, keeping the HTTPS
+protocol for the "git pull" and "git fetch" commands by issuing the following
+command:
+
+```
+git remote set-url --push origin git@github.com:input-output-hk/cardano-sl
+```
+
+### Listing the current and local branches
+```
+git branch
+```
+To list all the branches both local and at the origin (usually Github):
+```
+git branch --all
+```
+
+### Create and change into a new branch
+Development on Cardano SL is usually done by branching off the "develop" branch.
+Assuming you are already on the develop branch and your user name is "xyz" and
+you are working on YouTrack ticket number "YT-1234":
+```
+git branch xyz/YT-1234
+git checkout xyz/YT-1234
+```
+A shorter version of the above is:
+```
+git checkout -b xyz/YT-1234
+```
+
+### Rebasing your changes against develop
+While you are working on your branch, it is not uncommon for other developers
+to merge their changes into the develop branch. To pull those changes into
+your branch:
+```
+git fetch origin
+git rebase origin/develop
+```
+Occasionally when you rebase against "develop", you will get merge conflicts.
+Resolving merge conflicts is explained
+[here](https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/).
+
+### Editing the commit message of the most recent commit
+This is useful if you have just made a commit with just a commit message
+title and now want to add more text to the commit message body.
+```
+git commit --amend
+```
```

Comment:
I'll add: 

> Additionally, one can quickly add or remove a few changes from a last commit by doing:
> 
> ```
> git commit --amend -C HEAD
> ```
> 
> The `-C / --reuse-message` allows git to re-use the message of a previous commit. Passing `HEAD` as a reference here makes git apply the changes from the index to the last commit (`--amend`), re-using the message of that last commit (`-C HEAD`).  
> 
> ### Commit Surgery
>
> In order to amend commits further back in the history, the `-i / --interactive` option on a `rebase` command comes in handy. Git will prompt you with an editor allowing you to select a bunch of actions to apply amongst:
> 
> - `pick`: Do nothing and pick the commit as is.
> - `reword`: Change the commit's message without altering its content
> - `edit`: Edit the commit's message and its content (will require `git commit --amend`)
> - `squash`: Use commit, but meld into the previous one
> - `drop`: Remove commit
>
> Usually, you can run this command against your own branch, rebasing on a few commits before `HEAD` like the followings:
> 
> ```
> git rebase --preserve-merges -i HEAD~3
> ```
>
> Git will prompt you with a list of three commits and ask for actionit s. For instance, in order to reword the second commit and update the third one, one can do:
>
> ```
> pick 67aeecb Add PR checkbox for updating the CHANGELOG
> reword 834ade8 [DEVOPS-992] Fix x509 generator SANs to work with IP addresses
> edit ccbff81 [DEVOPS-980] Create build artifact with logs when test fails
> ```
>
> Git will re-play each commit, prompting you with an editor when you ask for a `reword` and, pausing the rebase on commits you asked to edit, right after having applied the relevant commit. From there, you can simply apply your changes and use `git commit --amend` as seen above to apply changes to the previous commit. Once done, `git rebase --continue` to move forward. 
>
> Note that, from the interactive rebase prompt, it is also possible to re-order commits as you wish. Squashing some of them together if needed. It is a really powerful command to restructure the history and make git tells the story you want it to tell!
>

Reply:
This was the "getting stated" guide which is a really good place to avoid getting too fancy.

Reply:
Hmmmm.. Then I'll need a place for this. Maybe confluence is the way to go. Or another guide about more complex git manipulations? I find it essential to be able to play with the commit history and analyze stuff this way. 
I've just completely rebuild a PR because all sort of commits were included in a rebase that weren't part of the initial PR :/ Running a `git rebase` shouldn't look magic for people, but be a rather natural thing to do. 
Netherless I agree this is maybe too advanced for a _Getting Started_ guide..

Reply:
I think the usage of `rebase -i` should be kept minimal as it requires a forced push. This is not usabel in a collaborative environment. But, I agree an interactive rebase with `fixup` is a nice tool in cleaning up a branch and for preparing a PR where only one person is the designated submitter.